### DISCUSSION TITLE
Allow common sum type fields and methods, subset conversion

### DISCUSSION THREAD

**Discussion by ntrel** [vlang/v#11693] [Ideas]
This code is very repetitive:
```v
pub fn (stmt Stmt) position() token.Position {
	match stmt {
		AssertStmt { return stmt.pos }
		AssignStmt { return stmt.pos }
		Comment { return stmt.pos }
		CompIf { return stmt.pos }
		ConstDecl { return stmt.pos }
		...
		else { return token.Position{} }
	}
```
Alex has a feature for a safer `sum as Type` that can't panic, requiring `if`:
```v
if sum is T {
	// typeof(sum) is T here
}
```
This could be extended to allow making `sum` a subset of its allowed types (I'm just showing 2 here):
```v
if stmt as AssertStmt | AssignStmt {
	// here stmt is known to only hold one of the above types
	return stmt.pos
}
return token.Position{}
```
`AssertStmt | AssignStmt` is a sum type literal. Each type must be listed in `Stmt` or it's a compile error.

`return stmt.pos` can work by lowering to:
```v
match stmt {
	AssertStmt { return stmt.pos }
	AssignStmt { return stmt.pos }
}
```
If one of the types doesn't have a field called `pos`, then the lowering would naturally fail to compile, using the existing code in `checker.v`.

Calling a method on a sum type instance can work with similar lowering.

**Comment by medvednikov** [comment]
Looks good, I'd keep using `is` for consistency

`if stmt is AssertStmt | AssignStmt {`

**Comment by spaceface777** [comment]
I would really like to see this feature implemented as well (I'm the one who proposed I in Discord :wink: ). I'll work on implementing it.

> I'd keep using `is` for consistency

I agree, that's the syntax I'll be using.

**Comment by ntrel** [comment]
> I'd keep using is for consistency

See https://github.com/vlang/v/issues/5826.
